### PR TITLE
Allow pass the color of points in 'dot-color' mode of Graph3D explicitly

### PR DESCRIPTION
**Pull request by ephraimberkovitch**
In Graph3D 'dot-color' mode, an addition option has been added - to explicitly set the data point color:

Now the color may be set as a string representation of HTML color, like #ff00ff,
or calculated from a number, for example, distance from this point
The first option is useful when we have some pre-given legend, to which we have to adjust ourselves
The second option is useful when we are interested in automatically setting the color, from some value,
using some color scale

### CODE REVIEW

**Comment by ephraimberkovitch** [comment]
Thanks, I will be glad to contribute to the library also in the future :)

**Comment by ephraimberkovitch** [comment]
Please, pay an attention to a new feature - onclick callback

File: examples/graph3d/07_dot_cloud_colors.html [hunk 1]
```diff
@@ -30,9 +33,13 @@
         var x = pow(random(), 2);
         var y = pow(random(), 2);
         var z = pow(random(), 2);
-        var dist = sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2));
+        var style;
```
Comment (mojoaxel) [inline examples/graph3d/07_dot_cloud_colors.html]: Maybe this is better written as: 
```js
var style = (i%2==0) ? sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2)) : "#00ffff";
```

File: lib/graph3d/Graph3d.js [hunk 2]
```diff
@@ -1996,6 +2011,7 @@ Graph3d.prototype._onMouseMove = function (event) {
   this.emit('cameraPositionChange', parameters);
 
   util.preventDefault(event);
+  //event.stopPropagation();
```
Comment (mojoaxel) [inline lib/graph3d/Graph3d.js]: we better do not include unused code

File: lib/graph3d/Graph3d.js [hunk 3]
```diff
@@ -2005,6 +2021,7 @@ Graph3d.prototype._onMouseMove = function (event) {
  * @param {event}  event   The event
  */
 Graph3d.prototype._onMouseUp = function (event) {
+  //this.moving = false;
```
Comment (mojoaxel) [inline lib/graph3d/Graph3d.js]: we better do not include unused code

File: lib/graph3d/Graph3d.js [hunk 4]
```diff
@@ -2015,6 +2032,22 @@ Graph3d.prototype._onMouseUp = function (event) {
...
+ */
+Graph3d.prototype._onClick = function (event) {
+  if (!this.moving) {
+      var boundingRect = this.frame.getBoundingClientRect();
+      var mouseX = getMouseX(event) - boundingRect.left;
+      var mouseY = getMouseY(event) - boundingRect.top;
+      var dataPoint = this._dataPointFromXY(mouseX, mouseY);
+      this.onclick_callback(dataPoint.point.data);
+  }
+  else // disable onclick callback, if it came immediately after rotate/pan
```
Comment (mojoaxel) [inline lib/graph3d/Graph3d.js]: please add the curly braces.

File: lib/graph3d/Graph3d.js [hunk 5]
```diff
@@ -1614,16 +1616,28 @@ Graph3d.prototype._getColorsRegular = function(point) {
...
-  var color       = this._hsv2rgb(hue, 1, 1);
-  var borderColor = this._hsv2rgb(hue, 1, 0.8);
+  var color, borderColor;
 
+  if (typeof point.point.value === "string") {
+    color = point.point.value;
+    borderColor = point.point.value;
+  }
+  else {
+      var hue         = (1 - (point.point.value - this.valueRange.min) * this.scale.value) * 240;
```
Comment (mojoaxel) [inline lib/graph3d/Graph3d.js]: it looks like the ident is not correct here.

File: lib/graph3d/Graph3d.js [hunk 6]
```diff
@@ -693,13 +693,15 @@ Graph3d.prototype.create = function () {
   var ontouchstart = function (event) {me._onTouchStart(event);};
   var onmousewheel = function (event) {me._onWheel(event);};
   var ontooltip = function (event) {me._onTooltip(event);};
+  var onclick = function(event) {me._onClick(event);}
```
Comment (mojoaxel) [inline lib/graph3d/Graph3d.js]: missing semicolon at the end of the line.

File: examples/graph3d/07_dot_cloud_colors.html [hunk 7]
```diff
@@ -13,6 +13,9 @@
     var data = null;
     var graph = null;
 
+    function onclick(point) {
+      console.log(point);
```
Comment (mojoaxel) [inline examples/graph3d/07_dot_cloud_colors.html]: sometimes `point` is `null` is you click outside the chart. This should be handled in the example.

Comment (ephraimberkovitch) [inline examples/graph3d/07_dot_cloud_colors.html]: In this case, it locates the closest point and returns it.

Comment (mojoaxel) [inline examples/graph3d/07_dot_cloud_colors.html]: There seems to be a problem than:
image

Comment (ephraimberkovitch) [inline examples/graph3d/07_dot_cloud_colors.html]: I have fixed this issue - sorry for not noticing that case before

**Comment by yotamberk** [comment]
@mojoaxel Can you review this again? The error has been solved

File: lib/graph3d/Graph3d.js [hunk 8]
```diff
@@ -956,52 +958,52 @@ Graph3d.prototype._dotSize = function() {
 
 
 /**
- * Get legend width
```
Comment (mojoaxel) [inline lib/graph3d/Graph3d.js]: Just a comment for the next time: Please do not change code-style of other people. This messes up the "git blame".

File: lib/graph3d/Graph3d.js [hunk 9]
```diff
@@ -2015,6 +2030,26 @@ Graph3d.prototype._onMouseUp = function (event) {
...
+ */
+Graph3d.prototype._onClick = function (event) {
+  if (!this.onclick_callback)
+    return;
+  if (!this.moving) {
+    var boundingRect = this.frame.getBoundingClientRect();
+    var mouseX = getMouseX(event) - boundingRect.left;
+    var mouseY = getMouseY(event) - boundingRect.top;
+    var dataPoint = this._dataPointFromXY(mouseX, mouseY);
+    if (dataPoint)
```
Comment (mojoaxel) [inline lib/graph3d/Graph3d.js]: As I said before. Please use curly brackets next time to improve readability and security.

File: lib/graph3d/Graph3d.js (hunk 8 above)
Comment (ephraimberkovitch) [inline lib/graph3d/Graph3d.js]: You are right, it was done automatically by my WebStorm, sorry for that

Comment (mojoaxel) [inline lib/graph3d/Graph3d.js]: Absolutely no problem! I actually had the same problem with my last commit. :-)

**Pull request merged**